config/intentions_service.go

```diff
@@ -0,0 +1,130 @@
+package config
+
+import (
+	"fmt"
+	"regexp"
+)
+
+// IntentionsServicesConfig configures regexp/list of names of services.
+type IntentionsServicesConfig struct {
+	// Regexp configures the source services to monitor by matching on the service name.
+	// Either Regexp or Names must be configured, not both.
+	Regexp *string `mapstructure:"regexp"`
+
+	// Names configures the services to monitor by listing the service name.
+	// Either Regexp or Names must be configured, not both.
+	Names []string `mapstructure:"names"`
+}
+
+// Copy returns a deep copy of IntentionsServicesConfig
+func (c *IntentionsServicesConfig) Copy() *IntentionsServicesConfig {
+	if c == nil {
+		return nil
+	}
+
+	var o IntentionsServicesConfig
+	o.Regexp = StringCopy(c.Regexp)
+
+	if c.Names != nil {
+		o.Names = make([]string, 0, len(c.Names))
+		o.Names = append(o.Names, c.Names...)
+	}
+	return &o
+}
+
+// Merge combines all values in this IntentionsServicesConfig with the values in the other
+// configuration, with values in the other configuration taking precedence.
+// Maps and slices are merged, most other values are overwritten. Complex
+// structs define their own merge functionality.
+func (c *IntentionsServicesConfig) Merge(o *IntentionsServicesConfig) *IntentionsServicesConfig {
+	if c == nil {
+		if o == nil {
+			return nil
+		}
+		return o.Copy()
+	}
+
+	if o == nil {
+		return c.Copy()
+	}
+
+	r := c.Copy()
+
+	if o.Regexp != nil {
+		r.Regexp = StringCopy(o.Regexp)
+	}
+
+	r.Names = mergeSlices(c.Names, o.Names)
+
+	return r
+}
+
+// Finalize ensures there no nil pointers.
+// Exception: `Regexp` is never finalized and can potentially be nil.
+func (c *IntentionsServicesConfig) Finalize() {
+	if c == nil { // config not required, return early
+		return
+	}
+
+	if c.Names == nil {
+		c.Names = []string{}
+	}
+}
+
+func (c *IntentionsServicesConfig) Validate() error {
+	if c == nil { // config not required, return early
```

Comment:
**config** not required?

Reply:
I believe that the nil check here is because the parent config struct could technically have a nil pointer to an undefined child struct. The parents recursively call Validate on their children. It's a pattern seen elsewhere in the code.

Reply:
As Derek explained this is a nil check for the parent config